libs/langchain/langchain/retrievers/pinecone_hybrid_search.py

```diff
@@ -1,12 +1,123 @@
 """Taken from: https://docs.pinecone.io/docs/hybrid-search"""
 
+from __future__ import annotations
+
 import hashlib
-from typing import Any, Dict, List, Optional
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
 
 from langchain.callbacks.manager import CallbackManagerForRetrieverRun
 from langchain.pydantic_v1 import Extra, root_validator
 from langchain.schema import BaseRetriever, Document
 from langchain.schema.embeddings import Embeddings
+from langchain.utils.iter import batch_iterate
+
+if TYPE_CHECKING:
+    from pinecone import Index
+from abc import ABCMeta, abstractmethod
+
+
+class PineconeIndexUpsert(metaclass=ABCMeta):
```

Comment:
@StankoKuveljic thanks for the contribution! 

Does the wrapper enable the user to do anything they can't already do when instantiating an Index for:


```
class PineconeHybridSearchRetriever(BaseRetriever):
    """`Pinecone Hybrid Search` retriever."""

    embeddings: Embeddings
    """Embeddings model to use."""
    """description"""
    sparse_encoder: Any
    """Sparse encoder to use."""
    index_upsert: PineconeIndexUpsert
```

----


For any run time arugments into upsert, we can definitely add them to add_texts

Reply:
Hey @eyurtsev thanks for the comments and review, I'll modify the code so there is no breaking change.

Short ansert is no.
However, there is a difference between upsert on single thread and upsert when `req_async=True` is passed on multithreaded pinecone index. Also, there will be more difference on how to upsert if we upgrade to `pinecone.GRPCIndex`.  Wanted to avoid having multiple if-else structure in `add_texts` method and use the same interface for upserting vectors regardless of which index  instance is used.

Reply:
What if we add a helper function called: `upsert` that takes an index and the various arguments `vectors`/`namespace`/batch_size`  and also parameters like `pool_threads` and `req_async`?

The conditional logic could be done in that function then, and I suspect it'll result in in less code and no intermediate abstractions.

This PR is also potentially adding hybrid search to the regular pinecone vectorstore -- would be good to hear your thoughts in terms of whether it'll work for your use case: https://github.com/langchain-ai/langchain/pull/13102/files

Reply:
> What if we add a helper function called: upsert that takes an index and the various arguments vectors/namespace/batch_size and also parameters likepool_threadsandreq_async`?

I wanted to avoid single function with conditional logic, but if you prefer that codestyle won't be problem to modify it.

So there is a plan to move hybridsearch implementation to regular Pinecone vector store? I wanted to suggest after my PR to refactor same pieces together into regular pinecone vectorstore package as most of the stuff is similar. I don't have an use-case, I'm working alongside Pinecone on langchain<>pinecone integration, so it might be better to ask community. As for the PR itself it seems good.
With this in mind, if the PR #13102 is going to be merged soon do we need my PR here? Will pinecone hybrid search be removed from retrievers and exist only in vectorstore?